azerothcore/azerothcore-wotlk
# fix(Core/Mails): Fix mails containing items not being returned if they were read and ensure mails sent by GMs are not returned

…eady returned before, and prevent mails sent by game masters from being returned once expired



## Changes Proposed:
-  Only delete expired mails containing items if they were already returned, regardless of being read or not by the receiver.
-  Always delete expired mails sent by game masters (GM mode on/console). Sometimes they don't even have a proper sender anyway (like when sent from console).
- Add a couple of helper functions to improve readibility.

## Issues Addressed:

- Closes https://github.com/chromiecraft/chromiecraft/issues/1857
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/8094

## SOURCE:

- TC Port https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Globals/ObjectMgr.cpp#L6374

## Tests Performed:


- Tested sending mails and having they returned.
- Tested sending mails with items and having they returned.
- Tested sending COD mails and having they returned.


## How to Test the Changes:


1. Send a mail with an item to some other character
2. Log that character and open the mail but don't take anything
3. Shut the server down
4. Use https://www.nirsoft.net/utils/run_as_date.html to time-travel and load your server as a date from the future (past the mail's expiration)
5. Notice that the mail was properly returned safe and sound with the item(s) you sent
6. Fast-forward again with Run as Date and check that the **returned** mail got deleted after it expired properly

## Known Issues and TODO List:



## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.
